python/link_prediction.py

```diff
@@ -0,0 +1,970 @@
+
+            # Process destination vertex next
+            dest_node = edge.to_vertex
+            dest_id, dest_type, dest_features = dest_node.id, merge_labels(dest_node.labels), dest_node.properties.get(link_prediction_parameters.node_features_property)
+ 
+            # Define type triplet
+            type_triplet = (src_type, edge_type, dest_type)
+            
+            type_triplet_in = type_triplet in type_triplets
+            
+            # Before processing node dest node and edge, check if this edge_type has occurred with different src_type or dest_type
+            if edge_type in edge_types and not type_triplet_in and edge_type == link_prediction_parameters.target_relation:
+                raise Exception(f"Edges of edge type {edge_type} are used for training and there are already edges with this edge type but with different combination of source and destination node. ")
+
+            # Add to the type triplets
+            if not type_triplet_in:
+                type_triplets.append(type_triplet)
+
+            # Add to the edge_types set
+            edge_types.add(edge_type)
+            
+            # Handle mappings
+            _process_help_function(mem_indexes, dest_id, dest_type, dest_features, reindex_orig, reindex_dgl, index_dgl_to_features) 
+
+            # Define edge
+            src_nodes[type_triplet].append(reindex_orig[src_type][src_id])
+            dest_nodes[type_triplet].append(reindex_orig[dest_type][dest_id])
+            
+            
+    # Check if there are no edges in the dataset, assume that it cannot learn effectively without edges. E2E handling.
+    if len(src_nodes.keys()) == 0:
+        raise Exception("No edges in the dataset. ")
+ 
+    # data_dict has specific type that DGL requires to create a heterograph 
+    data_dict = dict()  
+
+    # Create a heterograph
+    for type_triplet in type_triplets:
+        data_dict[type_triplet] = torch.tensor(src_nodes[type_triplet]), torch.tensor(dest_nodes[type_triplet])
+
+    g = dgl.heterograph(data_dict)  
+    
+    # Add self-loop support
+    transform = AddSelfLoop(allow_duplicate=False, new_etypes=True)
+    g = transform(g)
+
+    # Create features
+    for node_type in g.ntypes:
+        node_features = []
+        for node in g.nodes(node_type):
+            node_id = node.item()
+            node_features.append(index_dgl_to_features[node_type][node_id])
+
+        node_features = torch.tensor(node_features, dtype=torch.float32)
+        g.nodes[node_type].data[link_prediction_parameters.node_features_property] = node_features
+    
+    # Test conversion. Note: Do a conversion before you upscale features.
+    _conversion_to_dgl_test(graph=g, reindex_dgl=reindex_dgl, ctx=ctx, node_features_property=link_prediction_parameters.node_features_property)
+
+    # Upscale features so they are all of same size
+    _proj_0(g) 
+    
+    return g, reindex_dgl, reindex_orig
+
+def _validate_user_parameters(parameters: mgp.Map) -> None:
```

Comment:
This one is clearly for utility, it's long and repetitive, therefore -> hide it, bury it under ground :joy:!